EIPS/eip-4844.md

```diff
@@ -0,0 +1,558 @@
+The execution layer verifies the wrapper validity against the inner `TransactionPayload` after signature verification as:
+
+* All hashes in `blob_versioned_hashes` must start with the byte `BLOB_COMMITMENT_VERSION_KZG`
+* There may be at most `MAX_BLOBS_PER_TX` blob commitments in any single transaction.
+* There may be at most `MAX_BLOBS_PER_BLOCK` total blob commitments in a valid block.
+* There is an equal amount of versioned hashes, kzg commitments and blobs.
+* The KZG commitments hash to the versioned hashes, i.e. `kzg_to_versioned_hash(kzg[i]) == versioned_hash[i]`
+* The KZG commitments match the blob contents. (Note: this can be optimized with additional data, using a proof for a 
+  random evaluation at two points derived from the commitment and blob data)
+
+
+The signature is verified and `tx.origin` is calculated as follows:
+
+```python
+def tx_hash(tx: SignedBlobTransaction) -> Bytes32:
+    # The pre-image is prefixed with the transaction-type to avoid hash collisions with other tx hashers and types
+    return keccak256(BLOB_TX_TYPE + ssz.hash_tree_root(tx.message))
+
+def get_origin(tx: SignedBlobTransaction) -> Address:
+    sig = tx.signature
+    return ecrecover(tx_hash(tx), sig.v, sig.r, sig.s)
+```
+
+### Beacon chain validation
+
+On the consensus-layer the blobs are now referenced, but not fully encoded, in the beacon block body.
+Instead of embedding the full contents in the body, the contents of the blobs are propagated separately, as a "sidecar".
+
+This "sidecar" design provides forward compatibility for further data increases:
+with data-availability-sampling (DAS) not all blobs will have to be downloaded by all beacon nodes on the network.
+
+### BeaconBlockBody extension
+
+We add to the `BeaconBlockBody` a new object, `blob_kzgs: List[KZGCommitment, MAX_OBJECT_LIST_SIZE]`,
+and update `BeaconBlock` and `SignedBeaconBlock` to reflect this updated `BeaconBlockBody`.
+
+At the end of the `process_block` state-transition, the `BeaconBlock` and its contained `ExecutionPayload` 
+need to be cross-validated with `cross_validate(block.body)`.
+
+```python
+def cross_validate(body: BeaconBlockBody):
+    all_versioned_hashes = []
+    for tx in body.execution_payload.blob_transactions:
+        all_versioned_hashes.extend(tx.message.blob_versioned_hashes)
+    assert all_versioned_hashes == [
+        kzg_to_versioned_hash(kzg) for kzg in body.blob_kzgs
+    ]
+```
+
+### BlobsSidecar
+
+We introduce a `BlobsSidecar` container:
+
+```python
+class BlobsSidecar(Container):
+    beacon_block_root: Root
+    beacon_block_slot: Slot
+    blobs: List[Blob, MAX_OBJECT_LIST_SIZE]
+```
+
+Nodes broadcast the `SignedBeaconBlock` along with its corresponding
+`BlobsSidecar` over the network instead of plain beacon blocks.
+
+### Required blobs downloading
+
+When a node receives both objects, it first calls `verify_blobs` and if this call passes it runs the usual processing on the beacon block.
```

Comment:
Might consider using a function `is_data_available()` to abstractly talk about this verification. and then define it as all the data being in the sidecar for this EIP